### PR TITLE
Pass ESC Environments on stack update

### PR DESCRIPTION
**Pull request by seanyeh**
Related: https://github.com/pulumi/pulumi-service/pull/23672

We want to be able to track which ESC environments are depended on by which stacks on the cloud side, so this passes the list of imported envs as metadata on stack update

### CODE REVIEW

File: pkg/backend/updates.go
Comment (pgavlin) [inline pkg/backend/updates.go:32]: Feels like this should maybe be in `Environment` as another piece of metadata.

Comment (seanyeh) [reply pkg/backend/updates.go:32]: I just spoke with @dschaller about this and we will rename `EnvironmentImports` to `EscEnvironments` due to the confusion with `Environment` above. @pgavlin the `Environment` there is completely unrelated to ESC, does it still make sense to put ESC environments inside there?

Comment (pgavlin) [reply pkg/backend/updates.go:32]: I think so. Per the doc comment on `Environment`:

> Environment contains optional data from the deploying environment. e.g. the current source code control commit information.

To me, the ESC environment(s) in use feels like the same sort of info.

Comment (pgavlin) [reply pkg/backend/updates.go:32]: (cc @komalali for input here as well)

Comment (komalali) [reply pkg/backend/updates.go:32]: Yeah, I agree with @pgavlin - lets put it in Environment

Comment (seanyeh) [reply pkg/backend/updates.go:32]: @pgavlin @komalali i think conceptually that makes sense. since we're planning on storing ESC environments as structs (we want to store name and (eventually) resolved revision), is the play here to change `Environments` to be a `map[string]interface{}`? that comes with some annoying type checks, but if we want to stay backwards compatible I think that's the easiest?

Comment (dschaller) [reply pkg/backend/updates.go:32]: Isn't the environment here the execution environment? I feel like the esc environment is more configuration and less where it's running

Comment (pgavlin) [reply pkg/backend/updates.go:32]: I see two options:

1. Change this to a `map[string]any`. Requires changes in the service to accept the new type, but simplifies usage.
2. JSON-encode the list of environments and store that as the value for whatever key we use.

(1) is more work, but is probably preferable.

File: pkg/backend/backend.go
Comment (lunaris) [inline pkg/backend/backend.go:322]: I think at this point it'd probably be beneficial to add some doc comments here -- for people new to the codebase or unfamiliar with Pulumi, I think it'd otherwise be natural to assume that `Environment*` was talking about e.g. the OS environment.

**Review by lunaris — COMMENTED** [review]
It looks like this is going to change a bit before we should review it again -- is that right? If so feel free to ping me @seanyeh when it's ready for another look!

File: pkg/cmd/pulumi/import.go
Comment (seanyeh) [inline pkg/cmd/pulumi/import.go:907]: I have to reorder this so that we grab config before getting update metadata because we want to store the esc env information (retrieved from `getStackConfiguration`) into the metadata

File: pkg/cmd/pulumi/util.go
Comment (seanyeh) [inline pkg/cmd/pulumi/util.go:995]: turns out pat already added a key for this use case 🎉

File: pkg/backend/backend.go
Comment (komalali) [inline pkg/backend/backend.go:320]: Isn't `Environment` below also an ESC Environment? What's the distinction here? That's the resolved value and this is the list of imports?

File: pkg/cmd/pulumi/util.go
Comment (komalali) [inline pkg/cmd/pulumi/util.go:989]: Why not a `[]string` instead of `[]EscEnvironmentMetadata` since there's only 1 key?

**Review by pgavlin — APPROVED** [review]
LGTM pending responses to @komalali's feedback.

File: pkg/cmd/pulumi/util.go
Comment (seanyeh) [reply pkg/cmd/pulumi/util.go:989]: I talked with @dschaller about this and we would like to eventually add a `resolvedRevision` to this so we can track which revision was used. this will be a followup - I'll create an issue for this

File: pkg/backend/backend.go
Comment (seanyeh) [reply pkg/backend/backend.go:320]: Yah I can see that it's confusing. the `Environments` here is for the config values under `pulumiConfig` from the resolved esc values. I think it might be better if I rename this back to `EnvironmentImports` then

**Review by komalali — APPROVED** [review]

**Review by tgummerer — APPROVED** [review]

File: pkg/backend/backend.go
Comment (dschaller) [inline pkg/backend/backend.go:320]: Should we denote that these are ESC environment imports to differentiate between the environment values below and these

File: pkg/cmd/pulumi/util.go
Comment (dschaller) [inline pkg/cmd/pulumi/util.go:986]: Should we add a test for this?

File: pkg/backend/backend.go
Comment (dschaller) [reply pkg/backend/backend.go:320]: oh interesting - how does that work as a single value?

Comment (dschaller) [reply pkg/backend/backend.go:320]: synced offline - just confusion on my part about the names / pulumi config being able to be defined in the ESC env and also resolved as part of the CLI

**Review by dschaller — APPROVED** [review]

**Pull request merged**
